### PR TITLE
Fixing up the failures in the open integration tests.

### PR DESCRIPTION
**Pull request by tannergooding**
FYI. @dotnet/roslyn-infrastructure

### CODE REVIEW

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
Looks like there is still an issue where redeploying a VSIX won't cause the binding redirects to update.

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
So, it turns out, the binary formatter has some special logic for performing an assembly load when attempting to deserialize the return message of a call.

Depending on how an assembly was loaded the first time, it is possible that the call to `Assembly.Load()` will actually attempt to load the assembly, rather than resolving to the one that is already loaded in the `AppDomain`.

In this particular case, `Roslyn.VisualStudio.Test.Utilities` is loaded into both the server (VS) and client (xUnit) process. And even though the assemblies are the same/identical, they exist at two different paths (among other things, this means you can load both into a single AppDomain).

Anyways, the `HandleReturnMessage` call will end up calling `BinaryAssemblyInfo.GetAssembly()` which itself calls `Assembly.Load()` using the full name of the assembly that contains the serialized type.

To make a long story short, the two assemblies will be determined to be `'different'` and the client (xUnit) process will attempt to resolve the assembly.

In our scenario, the `BasePath` for the `AppDomain` will be in the NuGet packages cache and the assembly that it is trying to resolve will not be on the path, so the deserialization will fail.

So, I have added an `AssemblyResolver` which basically says, if an assembly matching the full name is already loaded into the `AppDomain` return that 😄

**Comment by tannergooding** [comment]
Explicitly tagging @jaredpar and @jasonmalinowski since I had to add an evil assembly resolver (see lengthy comment above detailing why this one is not evil).

**Comment by tannergooding** [comment]
Hmm... Tests are all passing, but not its getting version conflicts (and of course I updated no versions) 🙁

**Comment by tannergooding** [comment]
test open-vsi_prtest please
Version conflict, everything else passed.

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
test open-vsi_prtest please

Version conflict is due to differences in resolved assemblies for the VS 15 only build. Investigating further.

File: src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs [hunk 1]
```diff
@@ -122,7 +138,36 @@ public void AddProject(string projectName, string projectTemplate, string langua
         // TODO: Adjust language name based on whether we are using a web template
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: Is this comment now obsolete?

File: src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs [hunk 2]
```diff
@@ -122,7 +138,36 @@ public void AddProject(string projectName, string projectTemplate, string langua
...
+
+            switch (languageName.ToLower())
+            {
+                case "csharp":
+                {
+                    if (_csharpProjectTemplates.TryGetValue(projectTemplate, out projectTemplateName))
+                    {
+                        break;
+                    }
+                    goto default;
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: Oh goodness, let's not use this...that's really hard to read. Can we just make this a few if checks? It's half the vertical space and a lot less surprising.

File: src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs [hunk 3]
```diff
@@ -28,6 +28,24 @@ static VisualStudioInstanceFactory()
...
+            // Depending on the manner in which the assembly was originally loaded, this may end up actually trying to load the assembly a second
+            // time and it can fail if the standard assembly resolution logic fails. This ensures that we 'succeed' this secondary load by returning
+            // the assembly that is already loaded.
+            AppDomain.CurrentDomain.AssemblyResolve += (sender, eventArgs) => { 
+                Debug.WriteLine($"'{eventArgs.RequestingAssembly}' is attempting to resolve '{eventArgs.Name}'");
+                var resolvedAssembly = AppDomain.CurrentDomain.GetAssemblies().Where((assembly) => assembly.FullName.Equals(eventArgs.Name)).SingleOrDefault();
+
+                if (resolvedAssembly != null)
+                {
+                    Debug.WriteLine("The assembly was already loaded!");
```
Comment (jasonmalinowski) [inline src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs]: Isn't this going to be really chatty?

File: src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs (hunk 2 above)
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: Can do.

File: src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs (hunk 3 above)
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/VisualStudioInstanceFactory.cs]: No. By the time this gets added, most things have been resolved on the test side.

**Comment by tannergooding** [comment]
@jasonmalinowski, I talked with @jaredpar who just asked that I log a bug to move us to something like json-serialization that `shouldn't` have this problem.

It is trivial to produce a simple program that causes this (for example, one can just call `Assembly.Load` and `Assembly.LoadFrom`, where the assembly is the same but the path is different).

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
test open-vsi_prtest please

File: src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs (hunk 2 above)
Comment (tannergooding) [inline src/VisualStudio/TestUtilities/InProcess/SolutionExplorer_InProc.cs]: Fixed.

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Comment by tannergooding** [comment]
test open-vsi_prtest please

**Pull request merged**